src/backend/cdb/cdbmutate.c

```diff
@@ -1014,6 +1014,18 @@ shareinput_mutator_xslice_2(Node *node, PlannerInfo *root, bool fPop)
 					break;
 			}
 		}
+
+		/*
+		 * Remove writer gang for the shareinput scan readers in case
+		 * of cross-slice modifying CTE
+		 */
+		if (sisc->cross_slice &&
+			root->parse->hasModifyingCTE &&
+			currentSlice->gangType == GANGTYPE_PRIMARY_WRITER &&
+			!plan->lefttree)
+		{
+			currentSlice->gangType = GANGTYPE_PRIMARY_READER;
+		}
```

Comment:
Another moment is that `hadModifyingCTE` can relate to different CTE, and current SISC consumer can also ruin the gangType
**Without patch:**
```
create table t_repl (i int, j int) distributed replicated;
explain (slicetable,costs off) with cte as (insert into t_hashed select i, i*2 from generate_series(1,5) i  returning *),
cte2 as (select * from t_repl)
select * from cte join cte2 a on cte.i=a.i join (select t2.i from cte2 b join t2 using (i) limit 1) x on x.i=cte.i;
                                          QUERY PLAN                                           
-----------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)
   ->  Hash Join
         Hash Cond: (share0_ref2.i = t_hashed.i)
         ->  Shared Scan (share slice:id 1:0)
         ->  Hash
               ->  Hash Join
                     Hash Cond: (t_hashed.i = t2.i)
                     ->  Insert on t_hashed
                           ->  Redistribute Motion 1:3  (slice2; segments: 1)
                                 Hash Key: i.i
                                 ->  Function Scan on generate_series i
                     ->  Hash
                           ->  Broadcast Motion 1:3  (slice3; segments: 1)
                                 ->  Limit
                                       ->  Gather Motion 3:1  (slice4; segments: 3)
                                             ->  Limit
                                                   ->  Hash Join
                                                         Hash Cond: (share0_ref1.i = t2.i)
                                                         ->  Shared Scan (share slice:id 4:0)
                                                               ->  Seq Scan on t_repl t_repl_1
                                                         ->  Hash
                                                               ->  Seq Scan on t2
 Optimizer: Postgres-based planner
 Slice 0: Dispatcher; root 0; parent -1; gang size 0
 Slice 1: Primary Writer; root 0; parent 0; gang size 3
 Slice 2: Singleton Reader; root 0; parent 1; gang size 1; segment 1
 Slice 3: Singleton Reader; root 0; parent 1; gang size 1; segment 1
 Slice 4: Reader; root 0; parent 3; gang size 3
(28 rows)
```
With patch:
```
explain (slicetable,costs off) with cte as (insert into t_hashed select i, i*2 from generate_series(1,5) i  returning *),
cte2 as (select * from t_repl)
select * from cte join cte2 a on cte.i=a.i join (select t2.i from cte2 b join t2 using (i) limit 1) x on x.i=cte.i;
                                          QUERY PLAN                                           
-----------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)
   ->  Hash Join
         Hash Cond: (share0_ref2.i = t_hashed.i)
         ->  Shared Scan (share slice:id 1:0)
         ->  Hash
               ->  Hash Join
                     Hash Cond: (t_hashed.i = t2.i)
                     ->  Insert on t_hashed
                           ->  Redistribute Motion 1:3  (slice2; segments: 1)
                                 Hash Key: i.i
                                 ->  Function Scan on generate_series i
                     ->  Hash
                           ->  Broadcast Motion 1:3  (slice3; segments: 1)
                                 ->  Limit
                                       ->  Gather Motion 3:1  (slice4; segments: 3)
                                             ->  Limit
                                                   ->  Hash Join
                                                         Hash Cond: (share0_ref1.i = t2.i)
                                                         ->  Shared Scan (share slice:id 4:0)
                                                               ->  Seq Scan on t_repl t_repl_1
                                                         ->  Hash
                                                               ->  Seq Scan on t2
 Optimizer: Postgres-based planner
 Slice 0: Dispatcher; root 0; parent -1; gang size 0
 Slice 1: Reader; root 0; parent 0; gang size 3
 Slice 2: Singleton Reader; root 0; parent 1; gang size 1; segment 2
 Slice 3: Singleton Reader; root 0; parent 1; gang size 1; segment 2
 Slice 4: Reader; root 0; parent 3; gang size 3
(28 rows)

```